src/Nethermind/Nethermind.Consensus/Processing/CensorshipDetector.cs

```diff
@@ -0,0 +1,108 @@
+// SPDX-FileCopyrightText: 2024 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Nethermind.Blockchain;
+using Nethermind.Core;
+using Nethermind.Core.Caching;
+using Nethermind.Core.Collections;
+using Nethermind.Int256;
+using Nethermind.Logging;
+using Nethermind.TxPool;
+
+namespace Nethermind.Consensus.Processing;
+
+public class CensorshipDetector : IDisposable
+{
+    private readonly ITxPool _txPool;
+    private readonly IBlockProcessor _blockProcessor;
+    private readonly IBlockTree _blockTree;
+    private readonly IAccountStateProvider _accounts;
+    private readonly ILogger _logger;
+
+    private const int UnconfirmedBlocksCacheSize = 64;
+    private readonly LruCache<long, bool>
+        _unconfirmedBlocksCensorshipDetector = new(
+            UnconfirmedBlocksCacheSize, UnconfirmedBlocksCacheSize, "unconfirmedBlocksCensorshipDetector"
+            );
+    private const int CacheSize = 4;
+    private readonly LruCache<long, bool>
+        _censorshipDetector = new(CacheSize, CacheSize, "censorshipDetector");
+
+    public CensorshipDetector(
+        ITxPool txPool,
+        IBlockProcessor blockProcessor,
+        IBlockTree blockTree,
+        IAccountStateProvider accounts,
+        ILogManager logManager)
+    {
+        _txPool = txPool;
+        _blockProcessor = blockProcessor;
+        _blockTree = blockTree;
+        _accounts = accounts;
+        _logger = logManager?.GetClassLogger() ?? throw new ArgumentNullException(nameof(logManager));
+        _blockProcessor.BlockProcessing += OnBlockProcessing;
+        _blockTree.BlockAddedToMain += OnBlockAdded;
+    }
+
+    private void OnBlockProcessing(object? sender, BlockEventArgs e)
+    {
+        Task.Run(() => PreCachingOperation(e.Block));
+    }
+
+    private void OnBlockAdded(object? sender, BlockEventArgs e)
+    {
+        Task.Run(() => Cache(e.Block));
+    }
+
+    private void PreCachingOperation(Block block)
+    {
+        UInt256 maxGasPriceInBlock = 0;
+        Transaction[] blockTransactions = block.Transactions;
+        foreach (Transaction tx in blockTransactions)
+        {
+            if (!tx.SupportsBlobs)
+            {
+                if (tx.Supports1559)
+                {
+                    maxGasPriceInBlock = UInt256.Max(maxGasPriceInBlock, tx.GasPrice);
+                }
+                else
+                {
+                    maxGasPriceInBlock = UInt256.Max(maxGasPriceInBlock, tx.GasPrice + block.BaseFeePerGas);
+                }
```

Comment:
it should be in reverse: gas price + base fee for 1559 and just gas price for not supporting 1559

Reply:
```suggestion
                if (tx.Supports1559)
                {
                    UInt256 txGasPrice = UInt256.Min(tx.MaxFeePerGas, tx.MaxPriorityFeePerGas + block.BaseFeePerGas);
                    maxGasPriceInBlock = UInt256.Max(maxGasPriceInBlock, txGasPrice);
                }
                else
                {
                    maxGasPriceInBlock = UInt256.Max(maxGasPriceInBlock, tx.GasPrice);
                }
```
I am considering changing it according to the calculation given in the [GasPriceTxComparerHelper.cs](https://github.com/NethermindEth/nethermind/blob/master/src/Nethermind/Nethermind.Consensus/Comparers/GasPriceTxComparerHelper.cs) file

Reply:
Good idea, both would work.
More context: You don't need to check `MaxFeePerGas` - if tx was included, it means it was high enough. At this point only matters `MaxPriorityFeePerGas` and `BaseFee`. But if you use comparer, it will be actually even more simple - you can collect `bestTxInBlock` instead of `maxGasPriceInBlock` and then just use comparer to compare `bestTxInBlock` with `bestTxInPool`, all comparing-related logic would be reused, so you will be sure there is no mistake.

But in production code we use [GasPriceTxComparer](https://github.com/NethermindEth/nethermind/blob/master/src/Nethermind/Nethermind.Consensus/Comparers/GasPriceTxComparer.cs), `GasPriceTxComparerHelper` is used only if one of txs doesn't have calculated gasBottleneck and is called from `GasPriceTxComparer`